### PR TITLE
Adding OpenMPTarget parallel_scan

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

**Comment by crtrott** [comment]
Ignore all the removed code in OpenMP_Parallel.hpp that was just dead (commented) code before.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 1]
```diff
@@ -359,102 +359,103 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
 
   const FunctorType m_functor;
   const Policy m_policy;
-  /*
-    template< class TagType >
-    inline static
-    typename std::enable_if< std::is_same< TagType , void >::value >::type
-    exec_range( const FunctorType & functor
-              , const Member ibeg , const Member iend
-              , reference_type update , const bool final )
```
Comment (crtrott) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:373]: Note: all the removed code is irrelevant. That was dead (commented code) from the original CPU OpenMP backend).

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 2]
```diff
@@ -465,6 +466,51 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
   //----------------------------------------
 };
 
+template <class FunctorType, class ReturnType, class... Traits>
```
Comment (crtrott) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:469]: I hope you guys are proud how I avoided duplicated code :-)

**Review by dalg24 — CHANGES_REQUESTED** [review]
This looks good but you forgot work tags.  Obviously we are not testing for it...   I think we would have caught that if we enabled the sorting unit test in algotithm

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 3]
```diff
@@ -359,102 +359,103 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
-        #endif
-        #endif
-        for ( Member iwork = ibeg ; iwork < iend ; ++iwork ) {
-          functor( iwork , update , final );
+        int64_t local_offset = team_id * chunk_size;
+
+#pragma omp for
+        for (int i = 0; i < chunk_size; i++) {
+          int64_t idx = local_offset + i;
+          if (idx < N) a_functor(idx, element_values(team_id, i), false);
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:396]: Shouldn't `idx`	 be of type `Policy::index_type`?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 4]
```diff
@@ -359,102 +359,103 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
+        }
+#pragma omp barrier
+        if (omp_get_thread_num() == 0) {
+          value_type sum = 0;
+          for (int i = 0; i < chunk_size; i++) {
+            sum += element_values(team_id, i);
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:402]: Why isn't this `ValueJoin::join`?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp [hunk 5]
```diff
@@ -359,102 +359,103 @@ class ParallelScan<FunctorType, Kokkos::RangePolicy<Traits...>,
...
+            element_values(team_id, i) = sum;
```
Comment (masterleinad) [inline core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:403]: `ValueOps::copy`?

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp (hunk 4 above)
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp]: yeah I am doing that.

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp]: added that

File: core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp (hunk 5 above)
Comment (crtrott) [reply core/src/OpenMPTarget/Kokkos_OpenMPTarget_Parallel.hpp:421]: we don't support array type scans so I think its not needed.

**Comment by crtrott** [comment]
Added WorkTag too.

**Review by masterleinad — APPROVED** [review]
Looks reasonable to me.

**Review by dalg24 — APPROVED** [review]

**Pull request merged**